yt_dlp/extractor/zingmp3.py

```diff
@@ -51,45 +80,101 @@ def _extract_item(self, item, fatal):
             'formats': formats,
-            'thumbnail': item.get('thumbnail'),
+            'thumbnail': item.get('thumbnail') or item.get('thumbnailM'),
             'subtitles': subtitles,
             'duration': int_or_none(item.get('duration')),
             'track': title,
-            'artist': item.get('artists_names'),
+            'artist': item.get('artistsNames') or item.get('artists_names'),
             'album': album.get('name') or album.get('title'),
-            'album_artist': album.get('artists_names'),
+            'album_artist': album.get('artistsNames') or album.get('artists_names'),
         }
 
     def _real_extract(self, url):
-        page_id = self._match_id(url)
-        webpage = self._download_webpage(
-            url.replace('://zingmp3.vn/', '://mp3.zing.vn/'),
-            page_id, query={'play_song': 1})
-        data_path = self._search_regex(
-            r'data-xml="([^"]+)', webpage, 'data path')
-        return self._process_data(self._download_json(
-            'https://mp3.zing.vn/xhr' + data_path, page_id)['data'])
+        mobj = re.match(self._VALID_URL, url)
+        self.song_id = mobj.group('id')
+        self.type_url = mobj.group('type')
+
+        slug_api = ''
+        if self.type_url == 'bai-hat' or self.type_url == 'embed':
+            slug_api = SLUG_API.get('bai_hat')
+        elif self.type_url == 'video-clip':
+            slug_api = SLUG_API.get("video_clip")
+        elif self.type_url == 'album' or self.type_url == 'playlist':
+            slug_api = SLUG_API.get("playlist")
+
+        # Create api from slug_api with hash signature
+        api = self.get_api_with_signature(name_api=slug_api, param={
+            "ctime": TIMESTAMP,
+            "id": self.song_id})
+        info = self.download_json(api)
+        return self._process_data(info.get('data'))
+
+    def download_json(self, api):
+        # For the first time making the request, we need to make one more temp request to update the cookie
+        # and the next time we don't need to do it.
+        # But if download with cookieFile, so we don't need to make one more temp request
+        if self._downloader.params.get("cookiefile"):
+            self._IS_UPDATE_COOKIES = True
+
+        if not self._IS_UPDATE_COOKIES:
+            self._download_json(api, video_id=self.song_id, note='Update cookies for request')
+            self._IS_UPDATE_COOKIES = True
+
+        return self._download_json(api, video_id=self.song_id)
+
+    def get_api_with_signature(self, name_api, param):
+        if not name_api:
+            return
+
+        def get_hash256(string):
+            return hashlib.sha256(string.encode('utf-8')).hexdigest()
+
+        def get_hmac512(string):
+            return hmac.new(SECRET_KEY, string.encode('utf-8'), hashlib.sha512).hexdigest()
+
+        text = ""
+        for k, v in param.items():
+            text += f"{k}={v}"
+
+        # hash the query param by 256 algorithm
```

Comment:
> query param

Shouldn't key-value pairs be joined by `&`?

Reply:
@Lesmiscore It's is query param but It's also the text to hash with 256 algorithm `sha256 = get_hash256(text)`